File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/interview/P313_SuperUglyNumber.java
package com.dsa.leetcode.labuladong.interview;

import java.util.PriorityQueue;

/**
 * @since 2024/8/1 20:30
 */
public class P313_SuperUglyNumber {

    public int nthSuperUglyNumber(int n, int[] primes) {
        PriorityQueue<int[]> q = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        for (int prime : primes) {
            q.add(new int[]{prime, prime, 1});
        }
        int[] res = new int[n];
        res[0] = 1;
        int p = 1;
        while (p < n) {
            int[] poll = q.poll();
            int product = poll[0];
            int prime = poll[1];
            int index = poll[2];
            if (product != res[p - 1]) {
                res[p++] = product;
            }
            q.add(new int[]{res[index] * prime, prime, index + 1});
        }
        return res[n - 1];
    }

    public static void main(String[] args) {
        int n = 12;
        int[] primes = {2, 7, 13, 19};
        P313_SuperUglyNumber p313 = new P313_SuperUglyNumber();
        System.out.println(p313.nthSuperUglyNumber(n, primes));
    }

}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/interview/P313_SuperUglyNumber.py
import heapq
from typing import List


class Solution:
    def nthSuperUglyNumber(self, n: int, primes: List[int]) -> int:
        res = [1] * n
        p = 1
        q = []
        for prime in primes:
            heapq.heappush(q, (prime, prime, 1))
        while p < n:
            product, prime, index = heapq.heappop(q)
            if product != res[p - 1]:
                res[p] = product
                p += 1
            heapq.heappush(q, (res[index] * prime, prime, index + 1))
        return res[-1]


if __name__ == '__main__':
    n = 12
    primes = [2, 7, 13, 19]
    solu = Solution()
    print(solu.nthSuperUglyNumber(n, primes))
